lib/ansible/playbook/role/definition.py

```diff
@@ -150,56 +152,63 @@ def _load_role_path(self, role_name):
 
-        if role_tuple:
+        # if role_tuple:
             # we found it, stash collection data and return the name/path tuple
-            self._role_collection = role_tuple[2]
-            return role_tuple[0:2]
+            # self._role_collection = role_tuple[2]
+            # return role_tuple[0:2]
 
         # We didn't find a collection role, look in defined role paths
         # FUTURE: refactor this to be callable from internal so we can properly order
         # ansible.legacy searches with the collections keyword
 
         # we always start the search for roles in the base directory of the playbook
-        role_search_paths = [
-            os.path.join(self._loader.get_basedir(), u'roles'),
-        ]
+        # role_search_paths = [
+            # os.path.join(self._loader.get_basedir(), u'roles'),
+        # ]
 
         # also search in the configured roles path
-        if C.DEFAULT_ROLES_PATH:
-            role_search_paths.extend(C.DEFAULT_ROLES_PATH)
+        # if C.DEFAULT_ROLES_PATH:
+            # role_search_paths.extend(C.DEFAULT_ROLES_PATH)
 
         # next, append the roles basedir, if it was set, so we can
         # search relative to that directory for dependent roles
-        if self._role_basedir:
-            role_search_paths.append(self._role_basedir)
+        # if self._role_basedir:
+            # role_search_paths.append(self._role_basedir)
 
         # finally as a last resort we look in the current basedir as set
         # in the loader (which should be the playbook dir itself) but without
         # the roles/ dir appended
-        role_search_paths.append(self._loader.get_basedir())
+        # role_search_paths.append(self._loader.get_basedir())
 
         # now iterate through the possible paths and return the first one we find
-        for path in role_search_paths:
-            path = templar.template(path)
-            role_path = unfrackpath(os.path.join(path, role_name))
-            if self._loader.path_exists(role_path):
-                return (role_name, role_path)
+        # for path in role_search_paths:
+            # path = templar.template(path)
+            # role_path = unfrackpath(os.path.join(path, role_name))
+            # if self._loader.path_exists(role_path):
+                # return (role_name, role_path)
 
         # if not found elsewhere try to extract path from name
-        role_path = unfrackpath(role_name)
-        if self._loader.path_exists(role_path):
-            role_name = os.path.basename(role_name)
-            return (role_name, role_path)
+        # role_path = unfrackpath(role_name)
+        # if self._loader.path_exists(role_path):
+            # role_name = os.path.basename(role_name)
+            # return (role_name, role_path)
+
+        # searches = (self._collection_list or []) + role_search_paths
+
+        finder = AnsibleRoleFinder(self._loader.get_basedir(), self._role_basedir, templar)
```

Comment:
loader basdir might not be same as the dir of the current task file nor play nor playbook, as imports/includes can change it, also those should stack and not just use the last:

`ansibile-playbook play.yml`
where
`
 play.yml
roles/stuff/tasks/main.yml
other/plays.yml
other/roles/stuff2/tasks/main.yml
`
opens up diff behaviors depending on where/how you import a role.

Reply:
This matches exactly the usage of the original `RoleDefinition` code, so your suggestion to change it is a bit confusing to me.